src/shell.js

```diff
@@ -262,48 +262,26 @@ else if (ENVIRONMENT_IS_WEB || ENVIRONMENT_IS_WORKER) {
     Module['arguments'] = arguments;
   }
 
-  if (typeof console !== 'undefined') {
-    if (!Module['print']) Module['print'] = function shell_print(x) {
-      console.log(x);
-    };
-    if (!Module['printErr']) Module['printErr'] = function shell_printErr(x) {
-      console.warn(x);
-    };
-  } else {
-    // Probably a worker, and without console.log. We can do very little here...
-    var TRY_USE_DUMP = false;
-    if (!Module['print']) Module['print'] = (TRY_USE_DUMP && (typeof(dump) !== "undefined") ? (function(x) {
-      dump(x);
-    }) : (function(x) {
-      // self.postMessage(x); // enable this if you want stdout to be sent as messages
-    }));
-  }
-
-  if (typeof Module['setWindowTitle'] === 'undefined') {
-    Module['setWindowTitle'] = function(title) { document.title = title };
-  }
+  Module['setWindowTitle'] = function(title) { document.title = title };
 }
 else {
   // Unreachable because SHELL is dependent on the others
   throw new Error('Unknown runtime environment. Where are we?');
 }
 
-if (!Module['print']) {
-  Module['print'] = function(){};
-}
-if (!Module['printErr']) {
-  Module['printErr'] = Module['print'];
-}
-if (!Module['arguments']) {
-  Module['arguments'] = [];
-}
-if (!Module['thisProgram']) {
-  Module['thisProgram'] = './this.program';
-}
-if (!Module['quit']) {
-  Module['quit'] = function(status, toThrow) {
-    throw toThrow;
+if (typeof console !== 'undefined') {
+  Module['print'] = console.log;
+  Module['printErr'] = console.warn || console.log;
+} else {
+  Module['print'] = Module['printErr'] = function(){};
+#if LEGACY_VM_SUPPORT
```

Comment:
`LEGACY_VM_SUPPORT` is a bit vague - what is considered a legacy VM, and what does it enable? Can we directly name the feature something like `SUPPORT_OLD_SPIDERMONKEY_CONSOLE_PRINT` or similar? (I presume this is for old SpiderMonkey that did not have `console.log` but had `print`?)

Reply:
Not just spidermonkey, d8 also had print and printErr early on, and I think webkit had print but not printErr (but less sure about that, I tested less on it).

Honestly I think we can just drop this section, but I figured since `LEGACY_VM_SUPPORT` already exists, why not let it cover this as well? Basically anyone using a non up-to-date VM can use that and get backwards compatibility, but most people don't need to think about it.

Reply:
Oh now I see indeed we have a `LEGACY_VM_SUPPORT` feature already, missed that it already existed. I feel that's not the kind of way we'd want to feature test for compatibility features, since it's either "all legacy" or "no legacy", without documenting what the actual legacy feature(s) are and which versions of software need which legacy features. I think we'd rather remove `LEGACY_VM_SUPPORT` and replace it with feature-specific or client-specific #ifdefs (whichever is easier?) Something like that was the idea with #5554. Perhaps we could have `#ifdef SUPPORT_JS_VM_CONSOLE_PRINT` and `#ifdef POLYFILL_PRIMITIVE_MATH_OPS` or similar, so that we would not put multiple features behind the same flag, but have one define per feature? Then the name will illustrate what the actual features are.

For related reference:

 - SpiderMonkey documented `console.log` for SpiderMonkey 53, (which was released on 2017-04-19): https://developer.mozilla.org/en-US/docs/Mozilla/Projects/SpiderMonkey/Shell_global_objects$compare?locale=en-US&to=1151230&from=1004343

 - It looks like d8 does not yet have console.log, from https://bugs.chromium.org/p/v8/issues/detail?id=4986 . (or could be that bug report is out of date?)

 - From https://stackoverflow.com/questions/34914397/why-doesnt-console-log-work-in-the-jsc-environment-but-it-works-in-safaris-deb/34914888 and https://stackoverflow.com/questions/19649932/javascriptcore-console-log?noredirect=1&lq=1, it looks like `jsc` did not have console.log at least in May 2017.

Reply:
On the one hand, I agree that for commonly useful features we should have specific flags. But I see legacy support as different, it's rarely used, and it's code we are very close to just removing entirely.

For those reasons, a single flag for all legacy stuff seems ok to me. But if I'm in the minority here I don't feel strongly.

Reply:
> It looks like d8 does not yet have console.log, from https://bugs.chromium.org/p/v8/issues/detail?id=4986 . (or could be that bug report is out of date?)

The bug is out-dated. d8 does have console.log, along with other console methods. See https://cs.chromium.org/chromium/src/v8/src/d8-console.cc?l=59